Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -6331,6 +6332,88 @@ void Document::append_pending_fullscreen_change(PendingFullscreenEvent::Type typ
     m_pending_fullscreen_events.append(PendingFullscreenEvent { type, element });
 }
 
+// https://fullscreen.spec.whatwg.org/#fullscreen-an-element
+void Document::fullscreen_element_within_doc(GC::Ref<Element> element)
+{
+    auto const get_hide_until = [&](auto const& popover_list) {
+        return HTML::HTMLElement::topmost_popover_ancestor(element, popover_list, nullptr, HTML::IsPopover::No);
+    };
+
+    // 1. Let hideUntil be the result of running topmost popover ancestor given
+    // element, null, and false.
+    auto hide_until = get_hide_until(showing_hint_popover_list());
+
+    // Address a bug in the spec https://github.com/whatwg/fullscreen/issues/245
+    if (hide_until == nullptr)
+        hide_until = get_hide_until(showing_auto_popover_list());
+
+    // Our hide_all_popovers_until takes a variant. topmost_popover_ancestor produces a Ptr<HTMLElement>
+    Variant<GC::Ptr<HTML::HTMLElement>, GC::Ptr<Document>> hide_until_argument { hide_until };
+
+    // 2. If hideUntil is null, then set hideUntil to element’s node document.
+    if (hide_until == nullptr)
+        hide_until_argument = element->owner_document();
+
+    // 3. Run hide all popovers until given hideUntil, false, and true.
+    HTML::HTMLElement::hide_all_popovers_until(hide_until_argument, HTML::FocusPreviousElement::No, HTML::FireEvents::Yes);
+
+    // 4. Set element’s fullscreen flag.
+    element->set_fullscreen_flag(true);
+    // 5. Remove from the top layer immediately given element.
+    remove_an_element_from_the_top_layer_immediately(element);
+    // 6. Add to the top layer given element.
+    add_an_element_to_the_top_layer(element);
+    element->invalidate_style(StyleInvalidationReason::Fullscreen);
+}
+
+// https://fullscreen.spec.whatwg.org/#fullscreen-element
+GC::Ptr<Element> Document::fullscreen_element() const
+{
+    for (auto const& el : top_layer_elements().in_reverse()) {
+        if (el->is_fullscreen_element())
+            return el;
+    }
+    return nullptr;
+}
+
+// https://fullscreen.spec.whatwg.org/#dom-document-fullscreenelement
+GC::Ptr<Element> Document::fullscreen_element_for_bindings() const
+{
+    GC::Ptr<Element> fullscreen_element = this->fullscreen_element();
+
+    if (!fullscreen_element) {
+        return nullptr;
+    }
+
+    // 1. If this is a shadow root and its host is not connected, then return null.
+    // We're not a shadow root
```

Comment:
This is not spec text, please prefix it with `// AD-HOC:`. But why is step 1 in the algo then?

Reply:
It's pretty unfortunate that DocumentOrShadowRoot doesn't correspond to a C++ "FooMixin" class like the other interface mixins we have do, but fixing that design problem seems out of scope of this change.

Reply:
Yeah, if/when we land this PR, I could gladly take a look at making that happen, if you want to, see if I can make some headway? I'll gladly be the "owner" of this going forward and make sure it stays up to date.